diff --git a/src/gridftp_hdfs.c b/src/gridftp_hdfs.c
index 07cb125..0d264e9 100644
--- a/src/gridftp_hdfs.c
+++ b/src/gridftp_hdfs.c
@@ -30,6 +30,11 @@ char err_msg[MSG_SIZE];
 int local_io_block_size = 0;
 int local_io_count = 0;
 
+// global variable for username, filename and event type
+char gridftp_user_name[PATH_MAX];
+char gridftp_file_name[PATH_MAX];
+char gridftp_transfer_type[10];
+
 static globus_mutex_t g_hdfs_mutex;
 static pthread_t g_thread_id;
 static int g_thread_pipe_fd;
@@ -603,12 +610,15 @@ hdfs_start(
     strlength = strlength < 256 ? strlength  : 256;
     hdfs_handle->username = globus_malloc(sizeof(char)*strlength);
     if (hdfs_handle->username == NULL) {
+        gridftp_user_name[0] = '\0';
         finished_info.result = GLOBUS_FAILURE;
         globus_gridftp_server_operation_finished(
             op, GLOBUS_FAILURE, &finished_info);
         return;
     }
     strncpy(hdfs_handle->username, session_info->username, strlength);
+    // also copy username to global variable gridftp_user_name
+    strncpy(gridftp_user_name, session_info->username, strlength);
 
     // Pull configuration from environment.
     hdfs_handle->replicas = 3;
diff --git a/src/gridftp_hdfs_recv.c b/src/gridftp_hdfs_recv.c
index faa41f8..fb131bc 100644
--- a/src/gridftp_hdfs_recv.c
+++ b/src/gridftp_hdfs_recv.c
@@ -4,6 +4,10 @@
 
 #define ADVANCE_SLASHES(x) {while (x[0] == '/' && x[1] == '/') x++;}
 
+// extern global variable "gridftp_file_name"
+extern char gridftp_file_name[PATH_MAX];
+extern char gridftp_transfer_type[10];
+
 // Forward declarations of local functions
 static void
 hdfs_handle_write_op(
@@ -235,6 +239,9 @@ hdfs_recv(
 
     hdfs_handle->op = op;
     hdfs_handle->pathname = transfer_info->pathname;
+    // copy file name string to global variable
+    strncpy(gridftp_file_name, transfer_info->pathname, strlen(transfer_info->pathname));
+    strcpy(gridftp_transfer_type, "upload");
 
     if ((rc = prepare_handle(hdfs_handle)) != GLOBUS_SUCCESS) goto cleanup;
 
diff --git a/src/gridftp_hdfs_send.c b/src/gridftp_hdfs_send.c
index b5a9903..77c29d6 100644
--- a/src/gridftp_hdfs_send.c
+++ b/src/gridftp_hdfs_send.c
@@ -2,6 +2,9 @@
 #include "gridftp_hdfs.h"
 #include <syslog.h>
 
+extern char gridftp_file_name[PATH_MAX];
+extern char gridftp_transfer_type[10];
+
 // Forward declarations of local functions
 
 typedef struct hdfs_read_s {
@@ -95,6 +98,9 @@ hdfs_send(
     globus_mutex_lock(hdfs_handle->mutex);
     hdfs_handle->pathname = transfer_info->pathname;
 
+    strncpy(gridftp_file_name, transfer_info->pathname, strlen(transfer_info->pathname));
+    strcpy(gridftp_transfer_type, "download");
+
     ADVANCE_SLASHES(hdfs_handle->pathname)
     if (strncmp(hdfs_handle->pathname, hdfs_handle->mount_point, hdfs_handle->mount_point_len)==0) {
         hdfs_handle->pathname += hdfs_handle->mount_point_len;
