From 865e3f8fc6c73c19d83bb3563d7532b62e84d02c Mon Sep 17 00:00:00 2001
From: John Thiltges <jthiltges2@unl.edu>
Date: Thu, 21 Apr 2022 12:26:14 -0500
Subject: [PATCH] Reorder variables to address compiler warning

---
 src/XrdChecksum.hh | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/src/XrdChecksum.hh b/src/XrdChecksum.hh
index 9d9b79d..147be22 100644
--- a/src/XrdChecksum.hh
+++ b/src/XrdChecksum.hh
@@ -39,8 +39,8 @@ private:
     ChecksumState & operator=(ChecksumState const &);
 
     const unsigned m_digests;
-    uint32_t m_crc32;
     uint32_t m_cksum;
+    uint32_t m_crc32;
     uint32_t m_adler32;
 
     unsigned m_md5_length;
-- 
2.31.1

