--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -7,6 +7,8 @@ set( CMAKE_MODULE_PATH ${PROJECT_SOURCE_DIR}/cmake )
 
 find_package( Classad REQUIRED )
 
+include(CheckCXXCompilerFlag)
+
 find_package(PkgConfig)
 pkg_check_modules( GLOBUS_RSL REQUIRED globus-rsl )
 
@@ -14,6 +16,14 @@ if( CMAKE_COMPILER_IS_GNUCXX )
   set( CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -Wall -Werror" )
 endif( CMAKE_COMPILER_IS_GNUCXX )
 
+# Compiling against classad headers fail due to the usage of C++11
+# on RHEL7.  This CMake compiler test was taken from the HTCondor cmake.
+check_cxx_compiler_flag(-std=c++11 cxx_11)
+if (cxx_11)
+  set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -std=c++11")
+endif (cxx_11)
+
+
 string(REPLACE ";" " " _GLOBUS_RSL_CFLAGS_OTHER "${GLOBUS_RSL_CFLAGS_OTHER}" )
 
 link_directories ( ${GLOBUS_RSL_LIBRARY_DIRS} )
@@ -45,8 +55,10 @@ configure_file (
   "${CMAKE_CURRENT_BINARY_DIR}/config/01-ce-router-defaults.conf"
 )
 
-install(TARGETS eval_rsl LIBRARY DESTINATION "${CMAKE_INSTALL_LIBDIR}/condor")
-install(TARGETS classad_ce LIBRARY DESTINATION "${CMAKE_INSTALL_LIBDIR}/condor")
+set(LIB_INSTALL_DIR lib${LIB_SUFFIX})
+
+install(TARGETS eval_rsl LIBRARY DESTINATION "${CMAKE_INSTALL_PREFIX}/${LIB_INSTALL_DIR}/condor")
+install(TARGETS classad_ce LIBRARY DESTINATION "${CMAKE_INSTALL_PREFIX}/${LIB_INSTALL_DIR}/condor")
 
 install(PROGRAMS src/condor-ce DESTINATION ${SYSCONF_INSTALL_DIR}/rc.d/init.d)
 install(PROGRAMS src/condor_ce_env_bootstrap src/condor_ce_router_defaults src/osg-wrapper DESTINATION ${SHARE_INSTALL_PREFIX}/condor-ce)
