diff --git a/frontend/glideinFrontendElement.py b/frontend/glideinFrontendElement.py
index 732b093..f6a4ab5 100755
--- a/frontend/glideinFrontendElement.py
+++ b/frontend/glideinFrontendElement.py
@@ -371,7 +371,7 @@ class glideinFrontendElement:
 
         self.populate_status_dict_types()
         glideinFrontendLib.appendRealRunning(self.condorq_dict_running,
-                                             self.status_dict_types['Running']['dict'])
+                                             self.status_dict_types['Total']['dict'])
 
         # TODO: should IdleCores/RunningCores be commented here?
         self.stats['group'].logGlideins({
@@ -1414,7 +1414,7 @@ class glideinFrontendElement:
                 status_format_list = list(status_format_list) + list(self.x509_proxy_plugin.get_required_classad_attributes())
 
             # Consider multicore slots with free cpus/memory only
-            constraint = '(GLIDECLIENT_Name=?="%s.%s") && (%s)' % (self.frontend_name, self.group_name, mc_idle_constraint)
+            constraint = '(GLIDECLIENT_Name=?="%s.%s") && (%s)' % (self.frontend_name, self.group_name, "True")
             # use the main collector... all adds must go there
             status_dict = glideinFrontendLib.getCondorStatus(
                               [None],
