diff --git a/CMakeLists.txt b/CMakeLists.txt
index 47b2dba..b09b721 100644
--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -15,11 +15,11 @@ message ("Running CMake version ${CMAKE_VERSION}")
 # The version numbers
 #
 # DON'T DELETE
-## CVMFS_VERSION 2.4.0
+## CVMFS_VERSION 2.3.99.2.pre2
 #---------------------
 set (CernVM-FS_VERSION_MAJOR 2)
-set (CernVM-FS_VERSION_MINOR 4)
-set (CernVM-FS_VERSION_PATCH 0)
+set (CernVM-FS_VERSION_MINOR 3)
+set (CernVM-FS_VERSION_PATCH 99.2.pre2)
 set (CernVM-FS_VERSION_STRING "${CernVM-FS_VERSION_MAJOR}.${CernVM-FS_VERSION_MINOR}.${CernVM-FS_VERSION_PATCH}")
 
 #
