diff --git a/src/XrdHdfs.cc b/src/XrdHdfs.cc
index 5373567..695bc65 100644
--- a/src/XrdHdfs.cc
+++ b/src/XrdHdfs.cc
@@ -529,9 +529,19 @@ int XrdHdfsFile::Close(long long int *)
 {
    static const char *epname = "close";
 
-// Release the handle and return
+// If writing, force a sync of bytes all the way to the last data node disk.
 //
    int ret = XrdOssOK;
+   if (fh != NULL && m_nextoff) {
+       errno = 0;
+       if (-1 == hdfsHSync(m_fs, fh)) {
+         ret = XrdHdfsSys::Emsg(epname, error, errno, "hsync", fname);
+       }
+    }
+
+// Release the handle and return
+//
+   ret = XrdOssOK;
    if (fh != NULL  && hdfsCloseFile(m_fs, fh) != 0) {
       ret = XrdHdfsSys::Emsg(epname, error, errno, "close", fname);
    }
