commit 5f6013f46ed0e0f478dcfb44ea1ea3c949a7f04d
Author: Brian Bockelman <bbockelm@cse.unl.edu>
Date:   Mon Mar 27 16:24:25 2017 -0500

    Remove spurious warning on empty dirs.

diff --git a/src/XrdHdfs.cc b/src/XrdHdfs.cc
index b303e01..b5627b9 100644
--- a/src/XrdHdfs.cc
+++ b/src/XrdHdfs.cc
@@ -265,14 +265,14 @@ int XrdHdfsDirectory::StatRet(struct stat *buf)
     static const char *epname = "StatRet";
 #endif
 
-  if (!isopen) return -EBADF;
+   if (!isopen) return -EBADF;
 
-// Lock the directory and do any required tracing
-//
-  if (!dh)  {
-     XrdHdfsSys::Emsg(epname,error,EBADF,"read directory",fname);
-     return -EBADF;
-  }
+   // Check for a null directory handle; this will occur if this object is
+   // invalid, or if it is valid but is an empty directory.
+   if ((numEntries > 0) && !dh)  {
+      XrdHdfsSys::Emsg(epname,error,EBADF,"read directory",fname);
+      return -EBADF;
+   }
 
   m_stat_buf = buf;
 
