commit 42e1a4d218ba67d3eec0d254d6edbf38c7beb5a3
Author: Brian Bockelman <bbockelm@cse.unl.edu>
Date:   Mon Mar 27 20:28:50 2017 -0500

    Make sure to set errno to 0.

diff --git a/src/XrdHdfs.cc b/src/XrdHdfs.cc
index b5627b9..5ead286 100644
--- a/src/XrdHdfs.cc
+++ b/src/XrdHdfs.cc
@@ -181,6 +181,7 @@ int XrdHdfsDirectory::Opendir(const char *dir_path, XrdOucEnv & client)
 // Open the directory and get it's id
 // HDFS returns NULL but sets errno to 0 if the directory exists and is empty.
 //
+   errno = 0;
    if (!(dh = hdfsListDirectory(fs, fname, &numEntries)) && errno) {
       isopen = false;
       return (errno < 0) ? -EIO : -errno;
