From 2fb02a8d84bbe0e17fbf542afde54b74520c2565 Mon Sep 17 00:00:00 2001
From: John Thiltges <jthiltges2@unl.edu>
Date: Sun, 5 Mar 2017 19:58:57 -0600
Subject: [PATCH 1/2] Free bpxy in XrdSecProtocolgsi::Authenticate when using
 VOMS plugin

---
 src/XrdSecgsi/XrdSecProtocolgsi.cc | 1 +
 1 file changed, 1 insertion(+)

diff --git a/src/XrdSecgsi/XrdSecProtocolgsi.cc b/src/XrdSecgsi/XrdSecProtocolgsi.cc
index 7416e06..6cde2fd 100644
--- a/src/XrdSecgsi/XrdSecProtocolgsi.cc
+++ b/src/XrdSecgsi/XrdSecProtocolgsi.cc
@@ -1831,6 +1831,7 @@ int XrdSecProtocolgsi::Authenticate(XrdSecCredentials *cred,
                // PEM base64
                bpxy = (*X509ExportChain)(hs->Chain, true);
                bpxy->ToString(spxy);
+               delete bpxy;
                Entity.creds = strdup(spxy.c_str());
                Entity.credslen = spxy.length();
             } else {
-- 
2.7.4

