From 265651e8a1873114cb2793ccdf7fcce9808dd430 Mon Sep 17 00:00:00 2001
From: John Thiltges <jthiltges2@unl.edu>
Date: Mon, 13 Dec 2021 10:26:51 -0600
Subject: [PATCH] [XrdSciTokens] Modifying std::map invalidates iterator

---
 src/XrdSciTokens/XrdSciTokensAccess.cc | 6 ++++--
 1 file changed, 4 insertions(+), 2 deletions(-)

diff --git a/xrootd/src/XrdSciTokens/XrdSciTokensAccess.cc b/src/XrdSciTokens/XrdSciTokensAccess.cc
index 29dcfacfb..899c643ff 100644
--- a/xrootd/src/XrdSciTokens/XrdSciTokensAccess.cc
+++ b/xrootd/src/XrdSciTokens/XrdSciTokensAccess.cc
@@ -959,9 +959,11 @@ private:
         if (now <= m_next_clean) {return;}
         std::lock_guard<std::mutex> guard(m_mutex);
 
-        for (auto iter = m_map.begin(); iter != m_map.end(); iter++) {
+        for (auto iter = m_map.begin(); iter != m_map.end(); ) {
             if (iter->second->expired()) {
-                m_map.erase(iter);
+                iter = m_map.erase(iter);
+            } else {
+                ++iter;
             }
         }
         Reconfig();
-- 
2.31.1

