From c58d929f21530ab636bb19f6351dc286560d178a Mon Sep 17 00:00:00 2001
From: John Thiltges <jthiltges2@unl.edu>
Date: Sun, 5 Mar 2017 20:02:28 -0600
Subject: [PATCH 2/2] Free X509 context, stack, and store in
 XrdCryptosslX509VerifyChain

---
 src/XrdCrypto/XrdCryptosslAux.cc | 5 +++++
 1 file changed, 5 insertions(+)

diff --git a/src/XrdCrypto/XrdCryptosslAux.cc b/src/XrdCrypto/XrdCryptosslAux.cc
index 23e648b..fecccc9 100644
--- a/src/XrdCrypto/XrdCryptosslAux.cc
+++ b/src/XrdCrypto/XrdCryptosslAux.cc
@@ -195,6 +195,11 @@ bool XrdCryptosslX509VerifyChain(XrdCryptoX509Chain *chain, int &errcode)
    if (!verify_ok)
       errcode = gErrVerifyChain;
 
+   // Free context, stack, and store
+   X509_STORE_CTX_free(ctx);
+   sk_X509_pop_free(stk, X509_free);
+   X509_STORE_free(store);
+
    return verify_ok;
 }
 
-- 
2.7.4

