diff --git a/src/condor_utils/store_cred.cpp b/src/condor_utils/store_cred.cpp
index 13ada22..38197ca 100644
--- a/src/condor_utils/store_cred.cpp
+++ b/src/condor_utils/store_cred.cpp
@@ -33,6 +33,7 @@
 #include "secure_file.h"
 #include "condor_base64.h"
 #include "zkm_base64.h"
+#include "subsystem_info.h"
 
 static int code_store_cred(Stream *socket, char* &user, char* &pw, int &mode);
 
@@ -602,6 +603,14 @@ get_cred_handler(void *, int /*i*/, Stream *s)
 		goto bail_out;
 	}
 
+	// if we are the shadow, we should ignore the user sent over the socket
+	// and use the one supplied by the owner of the job being run.
+	// THIS CODE IS A HACK AND SHOULD PROBABLY BE REWRITTEN
+	if (get_mySubSystem()->isType(SUBSYSTEM_TYPE_SHADOW)) {
+		free(user);
+		user = strdup(get_user_loginname());
+	}
+
 	result = sock->code(domain);
 	if( !result ) {
 		dprintf(D_ALWAYS, "get_passwd_handler: Failed to recv domain.\n");
